components/script_plugins/unrooted_must_root.rs

```diff
@@ -139,97 +114,148 @@ impl<'a, 'tcx> LateLintPass<'a, 'tcx> for UnrootedPass {
+                        false
+                    } else if did.is_box() && in_new_function {
+                        // box in new() is okay
+                        false
+                    } else {
+                        true
+                    }
+                },
+                ty::TyParam(param_ty) => {
+                    let ty_param_def = cx.tcx.generics_of(self.def_id).type_param(&param_ty, cx.tcx);
+                    if cx.tcx.has_attr(ty_param_def.def_id, "must_root") {
+                        ret = true;
+                        false
+                    } else {
+                        true
+                    }
+                },
+                ty::TyRef(..) => false, // don't recurse down &ptrs
+                ty::TyRawPtr(..) => false, // don't recurse down *ptrs
+                ty::TyFnDef(..) | ty::TyFnPtr(_) => false,
+                _ => true
+            }
+        });
+        ret
+    }
 
-        fn require_rooted(cx: &LateContext, in_new_function: bool, subexpr: &hir::Expr) {
-            let ty = cx.tables.expr_ty(&subexpr);
-            if is_unrooted_ty(cx, ty, in_new_function) {
-                cx.span_lint(UNROOTED_MUST_ROOT,
-                             subexpr.span,
-                             &format!("Expression of type {:?} must be rooted", ty))
+    fn has_unrooted_generic_substs(&self, did: hir::def_id::DefId, substs: &ty::subst::Substs) -> bool {
+        let cx = self.late_cx;
+
+        let generics = cx.tcx.generics_of(did);
+        for ty_param_def in &generics.types {
+            // If type has `#[must_root]`, then it is ok to
+            // give it a must-root type, so just skip.
+            if cx.tcx.has_attr(ty_param_def.def_id, "must_root") {
+                continue;
             }
-        }
 
-        match expr.node {
-            // Trait casts from #[must_root] types are not allowed
-            hir::ExprCast(ref subexpr, _) => require_rooted(cx, self.in_new_function, &*subexpr),
-            // This catches assignments... the main point of this would be to catch mutable
-            // references to `JS<T>`.
-            // FIXME: Enable this? Triggers on certain kinds of uses of DomRefCell.
-            // hir::ExprAssign(_, ref rhs) => require_rooted(cx, self.in_new_function, &*rhs),
-            // This catches calls; basically, this enforces the constraint that only constructors
-            // can call other constructors.
-            // FIXME: Enable this? Currently triggers with constructs involving DomRefCell, and
-            // constructs like Vec<JS<T>> and RootedVec<JS<T>>.
-            // hir::ExprCall(..) if !self.in_new_function => {
-            //     require_rooted(cx, self.in_new_function, expr);
-            // }
-            _ => {
-                // TODO(pcwalton): Check generics with a whitelist of allowed generics.
+            let arg_ty = substs.type_at(ty_param_def.index as usize);
```

Comment:
Note that here we are using the `index` field; I advised you to do this intentionally because we are iterating over the generics from the child first, then up to the parent, which means we aren't walking the substs from 0..N, but really we are working sort of like `i..N` and then `0..i` where `i` is the number of generics in the parent. That seems fine.